unumpy/_multimethods.py

```diff
@@ -824,25 +804,442 @@ def nancumsum(a, axis=None, dtype=None, out=None):
+def percentile(
+    a,
+    q,
+    axis=None,
+    out=None,
+    overwrite_input=False,
+    interpolation="linear",
+    keepdims=False,
+):
+    return (a, q, mark_non_coercible(out))
+
+
+@create_numpy(_first2argreplacer)
+@all_of_type(ndarray)
+def nanpercentile(
+    a,
+    q,
+    axis=None,
+    out=None,
+    overwrite_input=False,
+    interpolation="linear",
+    keepdims=False,
+):
+    return (a, q, mark_non_coercible(out))
+
+
+@create_numpy(_first2argreplacer)
+@all_of_type(ndarray)
+def quantile(
+    a,
+    q,
+    axis=None,
+    out=None,
+    overwrite_input=False,
+    interpolation="linear",
+    keepdims=False,
+):
+    return (a, q, mark_non_coercible(out))
+
+
+@create_numpy(_first2argreplacer)
+@all_of_type(ndarray)
+def nanquantile(
+    a,
+    q,
+    axis=None,
+    out=None,
+    overwrite_input=False,
+    interpolation="linear",
+    keepdims=False,
+):
+    return (a, q, mark_non_coercible(out))
+
+
+def _ureduce(a, axis):
+    nd = ndim(a)
+
+    if axis is None:
+        a = ravel(a)
+        dims = (1,) * nd
+    else:
+        if not isinstance(axis, collections.abc.Sequence):
+            axis = (axis,)
+
+        axis = _normalize_axis(nd, axis)
+        unselected_axis = tuple(set(range(nd)) - set(axis))
+
+        dims = list(a.shape)
+        for ax in axis:
+            dims[ax] = 1
+        dims = tuple(dims)
+
+        a = transpose(a, unselected_axis + axis)
+        a = a.reshape(a.shape[: len(unselected_axis)] + (-1,))
+
+    return a, dims
+
+
+def _median_default(a, axis=None, out=None, overwrite_input=False, keepdims=False):
+    if overwrite_input:
+        return NotImplemented
+
+    a, dims = _ureduce(a, axis)
+
+    mask = any(isnan(a), axis=-1).reshape(a.shape[0:-1] + (1,))
+    a = where(mask, nan, a)
+
+    a = sort(a, axis=-1)
+
+    N = a.shape[-1]
+
+    indexer = [slice(None)] * ndim(a)
+    index = N // 2
+    if N % 2 == 0:
+        indexer[-1] = slice(index - 1, index + 1)
+    else:
+        indexer[-1] = slice(index, index + 1)
+    indexer = tuple(indexer)
+
+    a = mean(a[indexer], axis=-1)
+
+    if keepdims:
+        a = a.reshape(dims)
+
+    if out is None:
+        return a
+
+    if a.shape != out.shape:
+        raise ValueError("out parameter must have the same shape as the output")
+
+    copyto(out, a, casting="unsafe")
+    return out
+
+
+@create_numpy(_self_argreplacer, default=_median_default)
+@all_of_type(ndarray)
+def median(a, axis=None, out=None, overwrite_input=False, keepdims=False):
     return (a, mark_non_coercible(out))
 
 
-@create_numpy(_reduce_argreplacer)
+def _self_weights_argreplacer(args, kwargs, dispatchables):
+    def replacer(*args, weights=None, **kwargs):
+        return dispatchables[0:-1], dict(weights=dispatchables[-1], **kwargs)
+
+    return replacer(*args, **kwargs)
+
+
+def _average_default(a, axis=None, weights=None, returned=False):
+    if weights is None:
+        weights = ones(a.shape)
```

Comment:
```suggestion
        avg = mean(a, axis=axis)
        return (avg if not returned else
                (avg, full(avg.shape, avg.size // a.size)))
```

Reply:
Well spotted, I misunderstood that case.